README.md

python-mobilvest
===============
[![Code Health](https://landscape.io/github/ron8mcr/python-mobilvest/master/landscape.svg?style=flat)](https://landscape.io/github/ron8mcr/python-mobilvest/master)
[![Build Status](https://travis-ci.org/ron8mcr/python-mobilvest.svg?branch=master)](https://travis-ci.org/ron8mcr/python-mobilvest)
[![Coverage Status](https://coveralls.io/repos/ron8mcr/python-mobilvest/badge.svg?branch=master)](https://coveralls.io/r/ron8mcr/python-mobilvest?branch=master)

https://github.com/ron8mcr/python-mobilvest

What's that
-----------

*python-mobilvest is a application for working with API of [mobilvest.ru](http://mobilvest.ru) - service for sending SMS 

API documentation is availbe at [http://online.mobilvest.ru/api/](http://online.mobilvest.ru/api/) (need authorization)

Dependence
-----------

- `requests` # for http communication
- `responses` # for tests

Getting started
---------------
* Install python-mobilvest:

``pip install -e git+https://github.com/ron8mcr/python-mobilvest
``

Examples
--------
```python
import mobilvest
from time import sleep

# Create an instance of MobilVestAPI
# using login and api_key
# (you can find it out at http://online.mobilvest.ru/api/)
mapi = mobilvest.MobilVestApi(login='user',
    api_key='da39a3ee5e6b4b0d3255bfef95601890afd80709')

# check balance
response = mapi.get_balance()
balance = response['money']

# if we have enough money, send some sms
if balance > 0.5:
    phone = '79998887766'
    response = mapi.send_sms(phone=phone,
                             text='Hello from web!', sender='web.ru')
    # check the state of sms
    sms_id = response[0][phone]['id_sms']

    # Server can't process id_sms for so fast
    sleep(5)
    while True:
        try:
            response = mapi.get_status(sms_id)
            break
        except mobilvest.CantGetStatus:
            sleep(1)
            continue

    if response[sms_id]['status'] == 'deliver':
        print "Message delivered!"
    else:
        print "Message wasn't delivered: {}".format(response[sms_id]['status'])

```


CHANGELOG
v0.1
----
+ Class with all API methods of mobilvest.ru


mobilvest/test_mobilvest.py
#!/usr/bin/env python
# coding: UTF-8
import unittest
import responses
import mobilvest
import urlparse
import datetime
from hashlib import md5


class TestsMobilVest(unittest.TestCase):

    def setUp(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/timestamp.php',
                      body='1432359515')

        self.valid_api_key = '123'
        self.mapi = mobilvest.MobilVestApi('user', self.valid_api_key)

    @responses.activate
    def test_request_params(self):
        def check_request(request):
            params = urlparse.parse_qs(urlparse.urlparse(
                request.path_url).query)
            params = {k: params[k][0] for k in params}
            self.assertIn('login', params,
                          "Parameter 'login' wasn't passed")
            self.assertIn('timestamp', params,
                          "Parameter 'timestamp' wasn't passed")
            self.assertIn('signature', params,
                          "Parameter 'signature' wasn't passed")

            recieved_signature = params.pop('signature')
            params_as_string = ''.join(str(params[k]) for k in sorted(params))
            valid_signature = md5(
                params_as_string + self.valid_api_key).hexdigest()
            self.assertEqual(recieved_signature, valid_signature)
            return (200, {}, {})

        responses.add_callback(
            responses.GET, 'http://online.mobilvest.ru/get/balance.php',
            callback=check_request,
            content_type='application/json',
        )
        self.mapi.get_balance()

    @responses.activate
    def test_rise_exception(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/balance.php',
                      body='{"error": 7}')
        self.assertRaises(mobilvest.ServerResponsedWithError,
                          self.mapi.get_balance)

    @responses.activate
    def test_empty_response(self):
        """ Если ответ пустой, сервер возвращает код ошибки 19.
        Ожидается, что в таком случае просто вернется None,
        а исключение не будет брошено
        """
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/balance.php',
                      body='{"error": 19}')
        balance = self.mapi.get_balance()
        self.assertIsNone(balance)

    @responses.activate
    def test_get_balance(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/balance.php',
                      body='{"money" : "69573.1","currency" : "RUR"}')

        balance_json = self.mapi.get_balance()
        self.assertIn('money', balance_json)

    @responses.activate
    def test_get_base(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/base.php',
                      body='''{
                                  "125452": {
                                      "name": "Valuable clients",
                                      "time_birth": "12:00:00",
                                      "day_before": "0",
                                      "local_time": "1",
                                      "birth_sender": "",
                                      "birth_text": "",
                                      "on_birth": "0",
                                      "count": "2255",
                                      "pages": "23"
                                  }
                              }''')

        base_json = self.mapi.get_base()
        for base in base_json.values():
            self.assertIn('name', base)
            self.assertIn('count', base)

    @responses.activate
    def test_get_senders(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/senders.php',
                      body='{"smstest":"completed","smstest2":"completed"}')

        senders = self.mapi.get_senders()
        self.assertIsNotNone(senders)

    @responses.activate
    def test_get_phone(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/phone.php',
                      body='{"error": 19}')

        base_id = "125452"
        page = 1
        phones_json = self.mapi.get_phone(base_id, page)
        self.assertIsNone(phones_json)

    @responses.activate
    def test_get_status(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/status.php',
                      body='''
                            {
                                "4091297100348873330001" : "not_deliver"
                            }
                            ''')

        sms_id = "4091297100348873330001"
        statuses_json = self.mapi.get_status(sms_id)
        self.assertIn(sms_id, statuses_json)

    @responses.activate
    def test_get_multiple_statuses(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/status.php',
                      body='''
                            {
                                "4091297100348873330001" : "not_deliver",
                                "4091297100348880230003" : "not_deliver"
                            }
                            ''')

        sms_ids = ["4091297100348873330001", "4091297100348880230003"]
        statuses_json = self.mapi.get_status(sms_ids)
        for i in sms_ids:
            self.assertIn(i, statuses_json)

    @responses.activate
    def test_send_sms(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/send.php',
                      body='''
                        {
                            "79029134225": {
                                "error": "0",
                                "id_sms": "4092112510348380960001",
                                "cost": "0.5",
                                "count_sms": "1"
                            }
                        }''')

        phone = "79029134225"
        text = "Hello world!"
        sender = "web.web"
        result_json = self.mapi.send_sms(phone, text, sender)
        self.assertIn(phone, result_json)
        self.assertIn('id_sms', result_json[phone])

    @responses.activate
    def test_send_multiple_sms(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/send.php',
                      body='''
                        {
                            "79029134225": {
                                "error": "0",
                                "id_sms": "4092112510348380960001",
                                "cost": "0.5",
                                "count_sms": "1"
                            },
                            "79029134226": {
                                "error": "0",
                                "id_sms": "4092112510348380970001",
                                "cost": "0.5",
                                "count_sms": "1"
                            }
                        }''')

        phones = ["79029134225", "79029134226"]
        text = "Hello world!"
        sender = "web.web"
        result_json = self.mapi.send_sms(phones, text, sender)
        for p in phones:
            self.assertIn(p, result_json)

    @responses.activate
    def test_find_on_stop(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/find_on_stop.php',
                      body='{"error": 19}')

        # запрошенного номера нет в стоп-листе
        on_stop_json = self.mapi.find_on_stop('79324354123')
        self.assertIsNone(on_stop_json)

    @responses.activate
    def test_add_to_stop(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/add2stop.php',
                      body='{"id" : "4419373"}')

        response = self.mapi.add_to_stop('79324354123')
        self.assertIn('id', response)

    @responses.activate
    def test_get_template(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/template.php',
                      body="""{
                                "test": {
                                    "template": "text",
                                    "up_time": "2014-08-28 15:22:25"
                                }
                            }""")

        templates = self.mapi.get_template()
        for i in templates.values():
            self.assertIn('template', i)

    @responses.activate
    def test_add_template(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/add_template.php',
                      body='{"id" : "4419373"}')

        name = 'template_1'
        text = "Hello, World!"
        response = self.mapi.add_template(name, text)
        self.assertIn('id', response)

    @responses.activate
    def test_stat_by_month(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/stat_by_month.php',
                      body='{"error": 19}')

        # пустая статистика за текущий месяц
        now = datetime.datetime.now()
        response = self.mapi.stat_by_month(now)
        self.assertIsNone(response)

    @responses.activate
    def test_get_operator(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/operator.php',
                      body='{"operator" : "AT&T"}')

        phone = '7821345312'
        response = self.mapi.get_operator(phone)
        self.assertIn('operator', response)

    @responses.activate
    def test_get_incoming(self):
        responses.add(responses.GET,
                      'http://online.mobilvest.ru/get/incoming.php',
                      body='''
                            {
                                "5597": {
                                    "date": "2014-10-27 05:47:24",
                                    "sender": "79022754620",
                                    "prefix": "51632",
                                    "text": "51632 TEST"
                                }
                            }
                        ''')

        now = datetime.datetime.now()
        incomings = self.mapi.get_incoming(now)
        for i in incomings.values():
            self.assertIn('sender', i)
            self.assertIn('text', i)


mobilvest/mobilvest.py
#!/usr/bin/env python
# coding: UTF-8

import requests
from hashlib import md5


class ServerResponsedWithError(Exception):
    pass


class CantGetStatus(ServerResponsedWithError):
    pass


class MobilVestApi(object):
    ERRORS = {
        1: 'Не указана подпись',
        2: 'Не указан логин',
        3: 'Не указан текст',
        4: 'Не указан телефон',
        5: 'Не указан отправитель',
        6: 'Не корректная подпись',
        7: 'Не корректный логин',
        8: 'Не корректное имя отправителя',
        9: 'Не зарегистрированное имя отправителя',
        10: 'Не одобренное имя отправителя',
        11: 'В тексте содержатся запрещенные слова',
        12: 'Ошибка отправки СМС',
        13: 'Номер находится в стоп-листе. Отправка на этот номер запрещена',
        14: 'В запросе более 50 номеров',
        15: 'Не указана база',
        16: 'Не корректный номер',
        17: 'Не указаны ID СМС',
        18: 'Не получен статус',
        19: 'Пустой ответ',
        20: 'Номер уже существует',
        21: 'Отсутствует название',
        22: 'Шаблон уже существует',
        23: 'Не указан месяц (Формат: YYYY-MM)',
        24: 'Не указана временная метка',
        25: 'Ошибка доступа к базе',
        26: 'База не содержит номеров',
        27: 'Нет валидных номеров',
        28: 'Не указана начальная дата',
        29: 'Не указана конечная дата',
        30: 'Не указана дата (Формат: YYYY-MM-DD)'
    }
    BASE_URL = 'http://online.mobilvest.ru/get/'

    def __init__(self, login, api_key):
        self.login = login
        self.api_key = api_key

    def _list_to_str(self, lst):
        """ Преобразование списка к строке, где значения разделены запятой
        Такое требование API mobilvest к спискам
        """
        return ','.join(map(str, lst))

    def _get_server_timestamp(self):
        r = requests.get('http://online.mobilvest.ru/get/timestamp.php')
        return int(r.text)

    def _prepare_params(self, params):
        """
        Подготовка данных для запроса: вставка signature, timestamp, login
        params - словарь с параметрами запроса, не касающимися безопасности,
        логина, подписи и т. д.
        Документация:
        Подпись (параметр signature) - md5 хэш, который формируется
        следующим образом:
        Все параметры из запроса нужно отсортировать в алфавитном порядке
        в строку, в конец строки добавить API ключ. При этом
        последовательность параметров непосредственно в
        запросе не имеет значения.
        """
        result = params.copy()
        result['login'] = self.login
        result['timestamp'] = self._get_server_timestamp()
        result['return'] = 'json'
        params_as_string = ''.join(str(result[k]) for k in sorted(result))
        result['signature'] = md5(params_as_string + self.api_key).hexdigest()
        return result

    def _call_api(self, url, params):
        """Вызов одной из API-функций
        url - адрес php страницы
        params - параметры, не относящиеся к безопасности
                                            (login, timestamp, signature)
        возвращает ответ сервера в JSON
        """
        params = self._prepare_params(params)
        response = requests.get(self.BASE_URL + url, params=params).json()
        if not response:
            return None
        if 'error' in response:
            # почему-то пустой ответ у mobilvest - ошибка
            if response['error'] == 19:
                return None
            msg = "{}".format(self.ERRORS[response['error']])
            # в случае, если сервер ещё не обработал отправленное смс,
            # возвращается ошибка, которую необходимо отличать
            # от всех остальных
            if response['error'] == 18:
                raise CantGetStatus(msg)
            raise ServerResponsedWithError(msg)
        return response

    def get_balance(self):
        """ Запрос баланса
        Пример ответа:
        {"money" : "69573.1","currency" : "RUR"}
        """
        params = {}
        url = "balance.php"
        return self._call_api(url, params)

    def get_base(self):
        """Запрос списка баз
        Пример ответа:
        {
            "125452": {
                "name": "Valuable clients",
                "time_birth": "12:00:00",
                "day_before": "0",
                "local_time": "1",
                "birth_sender": "",
                "birth_text": "",
                "on_birth": "0",
                "count": "2255",
                "pages": "23"
            },
            "125453": {
                "name": "Partners",
                "time_birth": "12:00:00",
                "day_before": "0",
                "local_time": "1",
                "birth_sender": "",
                "birth_text": "",
                "on_birth": "0",
                "count": "2266",
                "pages": "23"
            },
        }
        """
        params = {}
        url = "base.php"
        return self._call_api(url, params)

    def get_senders(self):
        """Запрос списка отправителей
        Пример ответа:
        {"smstest":"completed","smstest2":"completed"}
        """
        params = {}
        url = "senders.php"
        return self._call_api(url, params)

    def get_phone(self, base, page):
        """ Запрос номеров из базы
        base - ID базы
        page - номер страницы
        пример ответа:
        {
            "79687931116": {
                "name": "",
                "last_name": "",
                "middle_name": "",
                "date_birth": "0000-00-00",
                "male": "",
                "note1": "",
                "note2": "",
                "region": "",
                "operator": ""
            },
            "79874617237": {
                "name": "",
                "last_name": "",
                "middle_name": "",
                "date_birth": "0000-00-00",
                "male": "",
                "note1": "",
                "note2": "",
                "region": "",
                "operator": ""
            }
        }
        """
        url = "phone.php"
        params = {'base': base, 'page': page}
        return self._call_api(url, params)

    def get_status(self, state):
        """ Запрос статусов
        state - ID статуса
        пример ответа:
        {
            "4091297100348873330001" : "not_deliver",
            "4091297100348880230003" : "not_deliver"
        }
        """
        url = "status.php"
        if isinstance(state, list):
            params = {'state': self._list_to_str(state)}
        else:
            params = {'state': state}
        return self._call_api(url, params)

    def send_sms(self, phone, text, sender):
        """ Отправка СМС
        phone - Один номер, или список номеров (не более 50 номеров)
        text - Текст СМС сообщения
        sender - Имя отправителя (одно из одобренных на вашем аккаунте)
        пример ответа:
        {
            "79029134225": {
                "error": "0",
                "id_sms": "4092112510348380960001",
                "cost": "0.5",
                "count_sms": "1"
            }
        }
        """
        url = "send.php"
        params = {'sender': sender, 'text': text}
        if isinstance(phone, list):
            params['phone'] = self._list_to_str(phone)
        else:
            params['phone'] = phone
        return self._call_api(url, params)

    def find_on_stop(self, phone):
        """ Поиск номера в стоп-листе
        phone - Искомый номер
        пример ответа:
        {"time_in" : "2014-08-29 11:07:43","description" : "descr"}
        """
        url = "find_on_stop.php"
        params = {'phone': phone}
        return self._call_api(url, params)

    def add_to_stop(self, phone):
        """ Добавление номера в стоп-лист
        phone - Номер который нужно добавить в стоп-лист
        пример ответа:
        {"id" : "4419373"}
        """
        url = "add2stop.php"
        params = {'phone': phone}
        return self._call_api(url, params)

    def get_template(self):
        """ Запрос списка шаблонов
        Пример ответа:
        {
            "test": "{"
            template ":"
            text ", "
            up_time ":"
            2014 - 08 - 28 15: 22: 25 "}",
            "test22": "{"
            template ":"
            testtt 1321321231321 ", "
            up_time ":"
            2014 - 08 - 28 15: 39: 07 "}"
        }
        """
        params = {}
        url = "template.php"
        return self._call_api(url, params)

    def add_template(self, name, text):
        """ Запрос списка шаблонов
        name - Название шаблона
        text - Текст шаблона
        Пример ответа:
        {"id" : "4419373"}
        """
        params = {'name': name, 'text': text}
        url = "add_template.php"
        return self._call_api(url, params)

    def stat_by_month(self, month):
        """ Общая статистика за месяц по дням
        month - datetime-объект, с установленными месяцем и годом
        Пример ответа:
        {
            "2014-08-01": {
                "deliver": {
                    "cost": "0.500",
                    "parts": "1"
                },
            "2014-08-06": {
                "deliver": {
                    "cost": "0.600",
                    "parts": "1"
                },
                "not_deliver": {
                    "cost": "0.600",
                    "parts": "1"
                }
        }
        """
        params = {'month': month.strftime("%Y-%m")}
        url = "stat_by_month.php"
        return self._call_api(url, params)

    def get_operator(self, phone):
        """ Запрос оператора по номеру
        phone - Номер абонента
        Пример ответа:
        {"operator" : "AT&T"}
        """
        params = {'phone': phone}
        url = "operator.php"
        return self._call_api(url, params)

    def get_incoming(self, date):
        """ Запрос входящих СМС
        date - datetime-объект
        Пример ответа:
        {{"5597" : {"date":"2014-10-27 05:47:24", "sender":"79029734720",
                    "prefix":"51632", "text":"51632 TEST"}}}
        """
        params = {'date': date.strftime("%Y-%m-%d")}
        url = "incoming.php"
        return self._call_api(url, params)
